### PR TITLE
Optimize MN lists cache

### LINKED ISSUE #3496: IBD Memory Leak
This issue was initially reported to me by @quantumexplorer, and I was able to replicate it.

Memory usage rapidly increases at a certain point during IBD and leads to IBD coming to a standstill only processing 1-3 blocks per second.

I ran this test on a DO $10 / month instance, which has 1 CPU, 2G ram and was running ubuntu 18.04. I did not have any added swap on this machine. 

during header sync, slow & consistent increase in memory usage.
800000 headers 460 M RAM used
100000  520 M
124000 620 M

During IBD
120000 Blocks 790M
220000 877M
580000 1022M
848000 1.09G
969000 1.23G
1080000 1.70G Rapid memory usage growth 40 blocks/second
1086200 1.84G
1086444 1.85G about 3 blocks/second
1086560 1.86G 1 block/second free shows only 61k free
1086728 57k free 1 block / second
Actually ran out of memory... Nothing will launch now. Had to force shutdown on Digital Ocean

On hard restart blocks reset to 1070000

### PR DESCRIPTION
**Pull request by UdjinM6**
We already optimized on-disk storage by keeping full snapshots for every 576th block only. This applies a similar logic to our in-memory cache which should significantly reduce memory consumption. ~Based on #3505 atm to avoid merge conflicts.~ Should (at least partially) fix #3496.

### CODE REVIEW

File: src/evo/deterministicmns.cpp [hunk 1]
```diff
@@ -790,7 +843,7 @@ bool CDeterministicMNManager::BuildNewListFromBlock(const CBlock& block, const C
         } else if (tx.nType == TRANSACTION_QUORUM_COMMITMENT) {
             llmq::CFinalCommitmentTxPayload qc;
             if (!GetTxPayload(tx, qc)) {
-                assert(false); // this should have been handled already
+                return _state.DoS(100, false, REJECT_INVALID, "bad-protx-payload");
```
Comment (thephez) [inline src/evo/deterministicmns.cpp:846]: Should this be?
```suggestion
                return _state.DoS(100, false, REJECT_INVALID, "bad-qctx-payload");
```

Comment (xdustinface) [reply src/evo/deterministicmns.cpp:846]: Good catch. I checked the other occurrences of and ended up with this one 997f2084ff67ce7dae4e08e886f90889139dfbce which fixes your suggestion and also this one from this PR

https://github.com/dashpay/dash/blob/4bfad3acfd8970d17070316eda18a48770a3d3ed/src/evo/cbtx.cpp#L198-L202

and this one which is not coming from this PR but its still wrong

https://github.com/dashpay/dash/blob/4bfad3acfd8970d17070316eda18a48770a3d3ed/src/llmq/quorums_blockprocessor.cpp#L327-L332

There was already a `bad-qc-payload` used in thats why i picked this naming https://github.com/dashpay/dash/blob/4bfad3acfd8970d17070316eda18a48770a3d3ed/src/llmq/quorums_commitment.cpp#L136-L139

Comment (UdjinM6) [reply src/evo/deterministicmns.cpp:846]: Yeah, this belongs to #3504 to actually. Cherry-picked there, thanks! 👍

**Comment by PastaPastaPasta** [comment]
Sync finished in a good amount of time. Overall looks good, maybe could be improved a bit, but overall ACK after rebase

**Pull request merged**